### PR TITLE
Allow filtering by PR label

### PR DESCRIPTION
**Pull request by rafiss**
This is useful since we have a convention of labelling PRs with things
like `backport-21.1.x`, but we don't have any tooling that actually ties
these labels back to whether or not the PR was backported.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/6)

### CODE REVIEW

**Comment by rafiss** [comment]
I haven't tested this!! Would appreciate any tips on how to run backboard locally.

File: state.go [hunk 1]
```diff
@@ -360,6 +367,10 @@ func (p *pr) MergedAt() string {
 	return "(unknown)"
 }
 
+func (p *pr) Labels() []string {
+	return p.labels
+}
+
```
Comment (rail) [inline state.go:373]: This method is unused, mind removing it?

```suggestion
```

File: state.go [hunk 2]
```diff
@@ -431,6 +442,22 @@ type queryer interface {
...
+	// in the future, but there isn't much harm in keeping it.
+	if len(pr.Labels) != 0 {
+		var labelCount int
+		err := q.QueryRow(`SELECT count(*) FROM pr_labels WHERE pr_id = $1`, pr.GetID()).Scan(&labelCount)
+		if err == sql.ErrNoRows {
+			return false, nil
+		} else if err != nil {
+			return false, err
+		}
+		if labelCount != len(pr.Labels) {
```
Comment (rail) [inline state.go:456]: What happens if we remove a label from a PR? Maybe do something similar to what we do for [pr_commits](https://github.com/cockroachdb/backboard/blob/1ddb8749804f2d87f6562b164f459224c0e407d8/state.go#L477)?

File: state.go [hunk 3]
```diff
@@ -492,6 +519,18 @@ func syncPR(ctx context.Context, db *sql.DB, repo *repo, pr *github.PullRequest)
 				return err
 			}
 		}
+		for _, l := range pr.Labels {
+			if l.GetName() == "" {
```
Comment (rail) [inline state.go:523]: I wonder if we should limit the amount of labels we add. The dropdown may become very busy :)
Maybe add filter out all labels that don't meet a fixed list of patterns aka allow list?

File: state.go [hunk 4]
```diff
@@ -431,6 +442,22 @@ type queryer interface {
 }
 
 func isPRUpToDate(ctx context.Context, q queryer, pr *github.PullRequest) (bool, error) {
+	// Support for PR labels was added later, so some PRs that are otherwise
```
Comment (rail) [inline state.go:445]: How does it behave when we merge this? Not sure if we can backfill labels for older PRs.

**Review by rail — CHANGES_REQUESTED** [review]
I have some comments and questions.

I'll prep a change for README.md with instructions how to run this locally, so you can iterate faster.

File: state.go (hunk 1 above)
Comment (rafiss) [reply state.go:373]: the `Number()`, `MergedAt()`, and `URL()` functions are unused too -- do you think I should remove those too? I made `Labels()` to follow the convention of making accessors for internal fields.

File: state.go [hunk 5]
```diff
@@ -56,6 +56,12 @@ CREATE TABLE IF NOT EXISTS pr_commits (
 	PRIMARY KEY (pr_id, sha)
 );
 
+CREATE TABLE IF NOT EXISTS pr_labels (
+    pr_id int REFERENCES prs,
+    label string,
+    PRIMARY KEY (pr_id, label)
+)
```
[the remark this answers is not in the document]
Comment (rafiss) [reply state.go:63]: fixed, thanks!

File: state.go (hunk 2 above)
Comment (rafiss) [reply state.go:456]: This if block is just meant to backfill things initially. This check doesn't replace the one below -- if a label is removed, the code below will still run, and we will see that the updatedAt times don't match, so the PR is not up to date.

File: state.go (hunk 4 above)
Comment (rafiss) [reply state.go:445]: The logic here is meant to backfill labels for older PRs. If we retrieve a PR from the github API and it has labels, then we do this extra check to see if we have stored any labels for the PR already. If we have not stored labels, then we will update the PRs. If we have stored labels, then we will still do the check below based on updatedAt

File: state.go (hunk 1 above)
Comment (rail) [reply state.go:373]: I'd say kill'em all! :) Less code less headache. We can add them back anytime when we want them to use, they are very simple getters.

File: state.go (hunk 3 above)
Comment (rafiss) [reply state.go:523]: That sounds good! The only labels we care about here start with `backport`, so I can add a filter for that.

File: state.go (hunk 1 above)
Comment (rafiss) [reply state.go:373]: Actually, those ones are used in the templating logic, my bad. I also updated the HTML template to display these Labels too, so it's clear that the filtering actually works.

Comment (rail) [reply state.go:373]: Ohhh, right. Completely forgot about the template!

File: state.go (hunk 4 above)
Comment (rail) [reply state.go:445]: It sounds sane to me. We should probably test how it behaves, when you deploy this change against the old db.

**Comment by rafiss** [comment]
I have made all the changes and it looked good to me locally. Ready for another look whenever you have the time

File: state.go (hunk 4 above)
Comment (rail) [reply state.go:461]: I ran it locally and everything looked good to me.

**Review by rail — APPROVED** [review]

**Review by rail — APPROVED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/6)* status: all files reviewed, 1 unresolved discussion (waiting on @rail)

___
*[state.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/backboard/6#-MYMWc24k_e5tM6o3Rbt-r1-63:-MYgYh7JA7_AnOwso7kI:b4c9lyr) ([raw file](https://github.com/cockroachdb/backboard/blob/63eb652dd83d8d08e6a56b705fa16fc83bb18534/state.go#L63)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

fixed, thanks!
</blockquote></details>

:thumbsup:

**Comment by rafiss** [comment]
Thanks for your help! Let me know what you think the next step should be to merge/deploy this. I see there's a `push.sh` so I can give that a shot, or I'm also happy to leave it to you.

**Comment by rail** [comment]
> Thanks for your help! Let me know what you think the next step should be to merge/deploy this. I see there's a `push.sh` so I can give that a shot, or I'm also happy to leave it to you.

Let me deploy it - want to verify the latest changes we applied to the process.

**Comment by rail** [comment]
I had to run `GRANT ALL on TABLE pr_labels TO backboard;` as `rail` user, because the `backboard` user had no `DELETE` privileges and what caused `2021-04-21T15:04:21.719560567Z2021/04/21 15:04:21 sync error: pq: user backboard does not have DELETE privilege on relation pr_labels`

**Pull request merged**